dds/DCPS/RTPS/Sedp.cpp

```diff
@@ -3107,6 +3107,44 @@ void Sedp::data_received(DCPS::MessageId message_id,
 }
 #endif
 
+//This is a function to unify the notifcation of liveliness within RTPS
+//The local participant map is checked for associated entities and then they are notified
+//of liveliness if their QoS is compatibile
+void
+Sedp::notify_liveliness(const ParticipantMessageData& pmd)
+{
+  const RepoId& guid = pmd.participantGuid;
+  RepoId guid_participant = guid;
+  guid_participant.entityId = ENTITYID_PARTICIPANT;
+  RepoId prefix = pmd.participantGuid;
+  prefix.entityId = EntityId_t(); // Clear the entityId so lower bound will work.
+  bool is_automatic = true;
+  ACE_DEBUG((LM_DEBUG, "(%P|%t) %T Sedp::notify_liveliness(const ParticipantMessageData& pmd), %d\n",
+             *(long*)(&pmd.participantGuid.entityId)));
+  if (PARTICIPANT_MESSAGE_DATA_KIND_MANUAL_LIVELINESS_UPDATE == pmd.participantGuid.entityId) {
+    is_automatic = false;
+  }
+
+  for (LocalSubscriptionMap::const_iterator sub_pos = local_subscriptions_.begin(),
+         sub_limit = local_subscriptions_.end();
+       sub_pos != sub_limit; ++sub_pos) {
+    const DCPS::RepoIdSet::const_iterator pos =
+      sub_pos->second.matched_endpoints_.lower_bound(prefix);
+    if (pos != sub_pos->second.matched_endpoints_.end() &&
+        DCPS::equal_guid_prefixes(*pos, prefix)) {
+      DCPS::DataReaderCallbacks_rch sl = sub_pos->second.subscription_.lock();
+      if (sl) {
+        if (sub_pos->second.qos_.liveliness.kind == ::DDS::AUTOMATIC_LIVELINESS_QOS && is_automatic) {
+          sl->signal_liveliness(guid_participant);
```

Comment:
`signal_liveliness` might not care about the entity so you may get away with not having `guid_participant`.  If it doesn't you may want to consider making it that way.

Reply:
It looks like it is using it.